### PR TITLE
fix(all): component reusage

### LINKED ISSUE #15484: [4.0.0-beta.7] [Ion-Select-Option] Displays old values when empty array
# Bug Report

**Ionic Info**
```
Ionic:

   ionic (Ionic CLI)          : 4.1.2
   Ionic Framework            : @ionic/angular 4.0.0-beta.7
   @angular-devkit/core       : 0.7.0-rc.3
   @angular-devkit/schematics : 0.7.0-rc.3
   @angular/cli               : 6.1.5
   @ionic/ng-toolkit          : 1.0.0
   @ionic/schematics-angular  : 1.0.1

Cordova:

   cordova (Cordova CLI) : 8.0.0
   Cordova Platforms     : android 7.0.0
   Cordova Plugins       : cordova-plugin-ionic-keyboard 2.1.2, cordova-plugin-ionic-webview 2.0.2, (and 4 other plugins)

System:

   Android SDK Tools : 26.1.1 (D:\dev\androidSDK)
   NodeJS            : v8.11.3 (D:\dev\nodejs\node.exe)
   npm               : 6.4.1
   OS                : Windows 7
```

**Describe the Bug**
When empty array is supplied to `*ngFor` in `ion-select-option` element then it displays old values

**Steps to Reproduce**
1. Select course two
2. Open instructors select (it is empty)
3. Select course one
4. Open instructors select (it contains two elements as expected)
5. Select course two
6. Open instructors select - here is the bug - it is not empty, contains values from step 4

**Related Code**
Simplified but it shows the bug
```
    <ion-item>
      <ion-label>Courses</ion-label>
      <ion-select (ionChange)="courseChanged($event)" interface="action-sheet">
        <ion-select-option *ngFor="let course of courses$ | async" [value]="course.id">
          {{course.name}}
        </ion-select-option>
      </ion-select>
    </ion-item>
    <ion-item>
      <ion-label>Instructors</ion-label>
      <ion-select interface="action-sheet">
        <ion-select-option *ngFor="let instructor of instructors$ | async" [value]="instructor.id">
          {{instructor.name}}
        </ion-select-option>
      </ion-select>
    </ion-item>
    <p>Here are displayed array values</p>
    <ion-item *ngFor="let instructor of instructors$ | async">
      {{instructor.name}}
    </ion-item>
```

```
	courseId$ = new BehaviorSubject<number>(0);
	courses$: Observable<any[]>;

	instructors$: Observable<any[]>;

	ngOnInit() {
		this.courses$ = this.getAllCourses();

		this.instructors$ = this.courseId$.pipe(
			switchMap(id => this.getInstructorsNamesByCourseId(id)),
			tap(console.log)
		);
	}

	setCourseId(id: number) {
		this.courseId$.next(id);
	}

	courseChanged(e: CustomEvent) {
		this.setCourseId(e.detail.value);
	}

	// api call
	getAllCourses(): Observable<any[]> {
		return of([
			{id: 1, name: 'Course one (2 instructors)'},
			{id: 2, name: 'Course two (empty)'},
		]);
	}

	// api call
	getInstructorsNamesByCourseId(id: number): Observable<any[]> {
		if (id === 1) {
			return of([
				{id: 1, name: 'Instructor One'},
				{id: 2, name: 'Instructor Two'},
			]);
		} else {
			return of([]);
		}
	}
```

**Expected Behavior**
Instructors select should always be empty when course selected item is (course two empty).
Correct values are always displayed below selects.

### LINKED ISSUE #15788: ion-select should re-evaluate .text when list options updated
# Bug Report

```
Ionic:

   ionic (Ionic CLI)          : 4.1.2 (/usr/local/lib/node_modules/ionic)
   Ionic Framework            : @ionic/angular 4.0.0-beta.12
   @angular-devkit/core       : 0.7.2
   @angular-devkit/schematics : 0.7.2
   @angular/cli               : 6.1.5
   @ionic/ng-toolkit          : 1.0.7
   @ionic/schematics-angular  : 1.0.5

Cordova:

   cordova (Cordova CLI) : 8.0.0
   Cordova Platforms     : android 7.1.1, browser 5.0.4, ios 4.5.5
   Cordova Plugins       : cordova-plugin-ionic-keyboard 2.1.2, cordova-plugin-ionic-webview 2.0.2, (and 7 other plugins)

System:

   Android SDK Tools : 26.1.1 (/Users/chrisc/Library/Android/sdk)
   ios-deploy        : 2.0.0
   ios-sim           : 7.0.0
   NodeJS            : v8.11.3 (/usr/local/bin/node)
   npm               : 6.4.1
   OS                : macOS High Sierra
   Xcode             : Xcode 10.0 Build version 10A255
```

**Describe the Bug**
If an ion-select has ion-select-options (ngFor) bound to an array which is then updated async, the select options update OK but the selected value shown in the UI is not. This is because ion-select only sets this.text when this.valueChanged fires, not when the related ion-select-options children are refreshed. 

https://github.com/ionic-team/ionic/blob/master/core/src/components/select/select.tsx#L123 

Workaround:
Binding [selectedText] to the required selected value works:
```
<ion-select [(ngModel)]="offer.growerNumber" name="growerNumber" okText="OK" cancelText="Dismiss" placeholder="Choose..." [selectedText]="offer.growerNumber">
            <ion-select-option *ngFor="let i of growerNumbers" [value]="i">{{i}}</ion-select-option>
          </ion-select>
```
**Expected Behavior**
When data source for ion-select-option list resolves via async call, the selected value should be shown in the UI.

**Additional Context**
https://forum.ionicframework.com/t/ionic-4-ion-select-and-change-detection/141643/2

### LINKED ISSUE #15879: [Ionic V4-beta.11] ion-select interface="action-sheet" action sheet list of buttons do not clear when no ion-select-options are added to ion-select
# Bug Report

**Ionic Info**
Run `ionic info` from a terminal/cmd prompt and paste the output below.

Ionic:

   ionic (Ionic CLI) : 4.1.2

System:

   NodeJS : v8.11.1
   npm    : 6.3.0
   OS     : Windows 7

I am creating stenciljs web component(s): @stencil/core: 0.12.4

**Describe the Bug**
I have one ion-select component that I want to hold different ion-select-options based on a previous ion-segment group selection. the selection of each ion-segment selection builds an array of items to pass to my stencil component. (the 'item' list gets built appropriately). If I select an ion-segment that creates a list of length > 0 the action-sheet is populated as expected, however If I then select an ion-segment that populates a list of length == 0 (meaning there are no items in the ion-select-options) when I select the dropdown I expected to see only a cancel option but the previous list is still there. 

**Steps to Reproduce**
step1: render the ion-select (interface="action-sheet") component with 3 ion-select-option components
step2: re-render the ion-select component with 0 ion-select-option components. 

**Here is my component**

```
import { Component, Element, Event, EventEmitter, Listen, Prop, State } from '@stencil/core';

@Component({
  tag: 'tracker-planner-select-item',
  styleUrl: 'tracker-planner-select-item.scss'
})
export class TrackerPlannerSelectItem {
  @Element() el: HTMLTrackerPlannerSelectItemElement;

  @Prop() id: string;
  @Prop() items: { key: string,
                   value: string
                 }[];
  @Prop() label: string;

  @State() _items: { key: string,
                     value: string
                   }[];

  @Event() trackerSelectChange: EventEmitter;

  @Listen('ionChange')
  onIonChange(event: CustomEvent) {
    event.preventDefault();
    event.stopPropagation();

    this.trackerSelectChange.emit({ ...event.detail, id: this.id });
  }

  componentWillUpdate() {
    // Trying to clean up action-sheet
    this._items = (this.items && this.items.length > 0) ? this.items : undefined;

    (this.el.querySelector('ion-select#' + this.id) as HTMLIonSelectElement).interfaceOptions = {
                                                                                                  header: this.label,
                                                                                                  subHeader: 'Select the lesson for this plan'
                                                                                                };
  }

  render() {
    return (
      <ion-item>
        <ion-label>{this.label}</ion-label>
        <ion-select name={this.id ? this.id : null} id={this.id ? this.id : null} interface="action-sheet" placeholder="Select One">
          { this._items
            ? this._items.map(element => <ion-select-option value={element.key}>{element.value}</ion-select-option>)
            : null
          }
        </ion-select>
      </ion-item>
    );
  }
}
```

**Expected Behavior**
I expected to see the correct array being populated even if previously there were say 10 <ion-select-options /> and then the component was updated with 0 <ion-select-options />

**Additional Context**
After looking through the generated HTML code there are no <ion-select-options /> generated after rendering the component back to an empty array, however, the previous action-sheet-buttons are still present so I am guessing that the action-sheet-buttons is not being generated properly if the new array of ion-select-options is 0.

### LINKED ISSUE #16364: [4.0.0-beta.16] ion-slides not working properly on fresh page load
# Bug Report

**Ionic Info**
Run `ionic info` from a terminal/cmd prompt and paste the output below.

```Ionic:

   ionic (Ionic CLI)             : 4.3.1 (C:\Users\***\AppData\Roaming\npm\node_modules\ionic)
   Ionic Framework               : @ionic/angular 4.0.0-beta.16
   @angular-devkit/build-angular : 0.10.6
   @angular-devkit/schematics    : 7.0.6
   @angular/cli                  : 7.0.6
   @ionic/angular-toolkit        : 1.2.0

System:

   NodeJS : v10.13.0 (C:\Program Files\nodejs\node.exe)
   npm    : 5.6.0
   OS     : Windows 10
```

**Describe the Bug**
Ionic Slides Component is not working when the page is refreshed.

**Steps to Reproduce**
1. ionic start tester tabs --type=angular
2. In home.html add Slides like so:
```<ion-content padding>
    <ion-slides>
        <ion-slide>
            <h1>Slide 1</h1>
        </ion-slide>
        <ion-slide>
            <h1>Slide 2</h1>
        </ion-slide>
        <ion-slide>
            <h1>Slide 3</h1>
        </ion-slide>
    </ion-slides>
</ion-content>
```
3. ionic serve
4. Refresh the page where the slides are on with F5 (without needing to navigate)
5. Slides component is broken (sliding not working etc.)

**Related Code**
See steps to reproduce

**Expected Behavior**
Slides component should still work as expected.

### LINKED ISSUE #16453: [4.0.0-beta.17] Conditional ion-select-option not updating
# Bug Report

**Ionic Info**
Run `ionic info` from a terminal/cmd prompt and paste the output below.

```
Ionic:

   ionic (Ionic CLI)             : 4.3.1 (/Users/rh/.nvm/versions/node/v8.11.4/lib/node_modules/ionic)
   Ionic Framework               : @ionic/angular 4.0.0-beta.16
   @angular-devkit/build-angular : 0.10.5
   @angular-devkit/schematics    : 7.0.5
   @angular/cli                  : 7.0.5
   @ionic/angular-toolkit        : not installed

Cordova:

   cordova (Cordova CLI) : 8.0.0
   Cordova Platforms     : android 7.1.2, ios 4.5.5
   Cordova Plugins       : cordova-plugin-ionic-keyboard 2.1.2, cordova-plugin-ionic-webview 2.1.0, (and 13 other plugins)

System:

   Android SDK Tools : 26.1.1 (/Users/rh/Library/Android/sdk)
   ios-deploy        : 1.9.2
   ios-sim           : 6.0.0
   NodeJS            : v8.11.4 (/Users/rh/.nvm/versions/node/v8.11.4/bin/node)
   npm               : 6.4.1
   OS                : macOS High Sierra
   Xcode             : Xcode 10.1 Build version 10B61
```

**Describe the Bug**
I have a select box in which the items contain a integer value. When this integer value is decreased below zero we try to hide the item with an *ngIf. However the item remains and the label of the ion-select does not update. This only happens when the limit of 0 is reached and the ngif is triggered. 

if is add a button that ngif's the entire ion-select and toggle it on and off then it does update.

**Steps to Reproduce**
1. Add several options to a ion-select
2. Add *ngIf to the options
3. Dynamicly change the value that your ngiffing to trigger de ngif with a button or something
4. open the selectbox
5. be amazed the value is still there

**Related Code**
` <ng-container *ngFor="let productCredit of productCredits">
            <ion-select-option *ngIf="productCredit.bookableCredit > 0" [value]="productCredit">{{ productCredit.label }}</ion-select-option>
        </ng-container>
<ion-button (tap)="changedBookableCreditToLessThanZero()">Push me<ion-button>`

**Expected Behavior**
The values in a ion-select should update visually, in my case the should dissappear because i ng-iffed them.

**Additional Context**
List any other information that is relevant to your issue. Stack traces, related issues, suggestions on how to fix, Stack Overflow links, forum links, screenshots, OS if applicable, etc.

### LINKED ISSUE #17344: [4.0.0] ion-select interface="action-sheet" not refreshing the items according to the updated model
# Bug Report

**Ionic version:**
**4.0.0**

**Current behavior:**
The **ion-select** action sheet correctly renders the initial list of items. But if an item is removed programmatically from the model, the **action-sheet** interface doesn't reflect that. That is, if you click on the dropdown to open the **action-sheet** interface, it still contains the deleted item in the list.

**Expected behavior:**
The **ion-select** action sheet interface does not display the deleted item from the model.

**Steps to reproduce:**
Let's say I have a select dropdown defined as below. When an item is removed by click on **Delete Color** button, the model **colors** get updated. But if we click on the dropdown, the **action-sheet** interface still shows the delete color.

***.html**

      <ion-select interface="action-sheet" [(ngModel)]="selectedColor">
            <ion-select-option *ngFor="let eachColor of colors;"
                    [value]="eachColor">
                 {{eachColor.label}}
            </ion-select-option>
      </ion-select>
      ...
      <ion-button (click)="deleteColor(selectedColor)">
            Delete Color
      </ion-button>
***.ts**

      colors = [{ id: 1, label: 'Red' },  
                { id: 2, label: 'Green' },  
                { id: 3, label: 'Blue' }];
      selectedColor = this.colors[0];
      
      ...
      deleteColor(color) {
         this.colors = this.colors.filter(eachColor => eachColor.id !== color.id);
         this.selectedColor = this.colors[0];
      }

**Other information:**

**Ionic info:** 

```
Ionic:

   ionic (Ionic CLI)             : 4.10.0
   Ionic Framework               : @ionic/angular 4.0.0
   @angular-devkit/build-angular : 0.12.3
   @angular-devkit/schematics    : 7.2.3
   @angular/cli                  : 7.2.3
   @ionic/angular-toolkit        : 1.3.0

Cordova:

   cordova (Cordova CLI) : 8.1.2 (cordova-lib@8.1.1)
   Cordova Platforms     : android 7.1.4, ios 4.5.5
   Cordova Plugins       : cordova-plugin-ionic-webview 2.3.2, (and 11 other plugins)

System:

   Android SDK Tools : 26.1.1
   ios-deploy        : 2.0.0
   NodeJS            : v8.11.3
   npm               : 5.6.0
   OS                : macOS Mojave
   Xcode             : Xcode 10.1 Build version 10B61

```

### LINKED ISSUE #17890: Can`t slide ion-item-option after reorder ion-item
# Bug Report

**Ionic version:**

[x] **4.x**

**Current behavior:**

If an `<ion-item-sliding>` is dragged down in `<ion-list>`, it is possible to slide `<ion-item-options>`. But if the item is dragged up in the list, sliding is no longer possible.

**Expected behavior:**

Sliding should still be possible if the item is dragged up in the list

**Steps to reproduce:**
See gif above

**Related code:**

```
<ion-list>
  <ion-reorder-group (ionItemReorder)="reorderConfigurationsList($event)" disabled="false">
    <ion-item-sliding *ngFor="let configuration of _configurationsList">
      <ion-item lines="full" color="white">
        <ion-label>
          <h2>{{configuration.name}}</h2>
          <ion-note>Created: {{ configuration.created | date: 'MM/dd/yy HH:mm' }}</ion-note>
        </ion-label>
        <ion-reorder slot="end"></ion-reorder>
      </ion-item>
      <ion-item-options side="start">
        <ion-item-option>Favorite</ion-item-option>
        <ion-item-option color="danger">Share</ion-item-option>
      </ion-item-options>
      <ion-item-options side="end">
        <ion-item-option>Unread</ion-item-option>
      </ion-item-options>
    </ion-item-sliding>
  </ion-reorder-group>
</ion-list>
```
````
reorderConfigurationsList(ev) {
    const itemToMove = this._configurationsList.splice(ev.detail.from, 1 ) [0];
    this._configurationsList.splice(ev.detail.to, 0, itemToMove );
    ev.detail.complete();
  }
````
**Other information:**

**Ionic info:**

```
Ionic:

   ionic (Ionic CLI)             : 4.12.0 (/usr/local/lib/node_modules/ionic)
   Ionic Framework               : @ionic/angular 4.1.2
   @angular-devkit/build-angular : 0.13.6
   @angular-devkit/schematics    : 7.2.4
   @angular/cli                  : 7.3.6
   @ionic/angular-toolkit        : 1.4.1

Cordova:

   cordova (Cordova CLI) : 8.1.2 (cordova-lib@8.1.1)
   Cordova Platforms     : none
   Cordova Plugins       : no whitelisted plugins (0 plugins total)

System:

   ios-sim : 8.0.0
   NodeJS  : v10.15.3 (/usr/local/bin/node)
   npm     : 6.9.0
   OS      : macOS Mojave
   Xcode   : Xcode 10.2 Build version 10E125

```

### LINKED ISSUE #18843: bug: IOS ion-select initialised off-screen, first time a value is selected the value is not rendered until something changes in page.
# Bug Report

**Ionic version:**
[x] **4.x**

**Current behavior:**
This could very well be due to Apple's implementation of their WebView but it's only the ion-select elements I've noticed it on.

on iOS, found on device and in simulator on OSX;
If a page is taller than the height of the device screen and an ion-select element is positioned off-screen, below the bottom, when scrolling to the item and selecting an option for the first time (e.g. changing from empty or no value to a value) the chosen value can be seen briefly and is then not rendered.
If any styling or content is changed that causes the value to be redrawn (presumably), such as;
 - scrolling up, so the element is off-screen, then scrolling back down to bring it into view.
 - choosing a different value in the same ion-select
 - choosing a value in another select (not sure what this changes to affect the initial ion-select)
Then the value is rendered.

**Steps to reproduce:**
 - build a page with a height taller than the device screen height, and position an ion-select such that you must scroll to bring it into view from its initial position.
 - Select an option for that ion-select
 - See the value be set then disappear
 - Edit the same ion-select to a different option, or scroll it out of view and back into view.
 - See the value now being rendered correctly.

**Other information:**
Using safari's inspector to view the elements as they change shows nothing to be wrong, and something as simple as scrolling the ion-select element out and back into view will cause it to be redrawn which suggests it may be down to the WebView and/or an order of operations on updating elements and styles.

**Ionic info:**
```
Ionic:

   Ionic CLI                     : 5.2.3 (/usr/local/lib/node_modules/ionic)
   Ionic Framework               : @ionic/angular 4.5.0
   @angular-devkit/build-angular : 0.13.9
   @angular-devkit/schematics    : 7.3.9
   @angular/cli                  : 7.3.9
   @ionic/angular-toolkit        : 1.5.1

Capacitor:

   Capacitor CLI   : 1.0.0
   @capacitor/core : 1.0.0

Cordova:

   Cordova CLI       : 9.0.0 (cordova-lib@9.0.1)
   Cordova Platforms : android 8.0.0, ios 5.0.1
   Cordova Plugins   : cordova-plugin-ionic-keyboard 2.1.3, cordova-plugin-ionic-webview 4.1.0, (and 8 other plugins)

Utility:

   cordova-res : 0.5.1 (update available: 0.6.0)
   native-run  : 0.2.7 (update available: 0.2.8)

System:

   ios-deploy : 1.9.4
   ios-sim    : 8.0.1
   NodeJS     : v12.5.0 (/usr/local/bin/node)
   npm        : 6.10.1
   OS         : macOS Mojave
   Xcode      : Xcode 10.2.1 Build version 10E1001

```
(Capacitor is there, but I'm not making use of it...)

### PR DESCRIPTION
**Pull request by manucorporat**
fixes #18843
fixes #17344
fixes #16453
fixes #15879
fixes #15788 
fixes #15484
fixes #17890
fixes #16364

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [ ] Build (`npm run build`) was run locally and any changes were pushed
- [ ] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: core/src/components/menu/menu.tsx [hunk 1]
```diff
@@ -170,6 +170,8 @@ export class Menu implements ComponentInterface, MenuI {
     // register this menu with the app's menu controller
     menuCtrl!._register(this);
 
+    this.updateState();
```
Comment (manucorporat) [inline core/src/components/menu/menu.tsx]: i might need to move back to didLoad(), will test

File: core/src/components/picker-column/picker-column.tsx [hunk 2]
```diff
@@ -81,14 +71,23 @@ export class PickerColumnCmp implements ComponentInterface {
...
-
     this.tmrId = setTimeout(() => {
       this.noAnimate = false;
       this.refresh(true);
     }, 250);
   }
 
-  componentDidUnload() {
+  componentDidLoad() {
+    // get the height of one option
```
Comment (manucorporat) [inline core/src/components/picker-column/picker-column.tsx]: add comment of why this needs t be here:
// It needs to perform a DOM read after the first render

File: core/src/components/range/range.tsx [hunk 3]
```diff
@@ -188,10 +182,15 @@ export class Range implements ComponentInterface {
     this.gesture.setDisabled(this.disabled);
   }
 
-  componentDidUnload() {
+  connectedCallback() {
```
Comment (manucorporat) [inline core/src/components/range/range.tsx]: move before componentWillLoad()

File: core/src/components/textarea/textarea.tsx [hunk 4]
```diff
@@ -169,28 +169,22 @@ export class Textarea implements ComponentInterface {
...
-   * Emitted when the input has been removed.
-   * @internal
-   */
-  @Event() ionInputDidUnload!: EventEmitter<void>;
-
-  componentWillLoad() {
+  connectedCallback() {
     this.emitStyle();
+    this.debounceChanged();
+    document.dispatchEvent(new CustomEvent('ionInputDidLoad', {
```
Comment (adamdbradley) [inline core/src/components/textarea/textarea.tsx]: Can this get changed to use `this.el` instead? Stencil would be able to run with `document` just fine, but I'd be worried about other systems choking on it.

**Comment by brandyscarney** [comment]
Ran into a couple of issues while testing:

### Reorder

I am getting an error on the item/reorder test: http://localhost:3333/src/components/item/test/reorder

```
reorder:95 Uncaught ReferenceError: indexes is not defined
```

### Router

http://localhost:3333/src/components/router/test/basic

This test has a blank screen ^

File: core/src/components/tabs/tabs.tsx [hunk 5]
```diff
@@ -44,9 +44,10 @@ export class Tabs implements NavOutlet {
     if (!this.useRouter) {
       this.useRouter = !!document.querySelector('ion-router') && !this.el.closest('[no-router]');
     }
-    const tabs = this.tabs;
-    await this.select(tabs[0]);
-
+    if (!this.useRouter) {
```
Comment (brandyscarney) [inline core/src/components/tabs/tabs.tsx]: Couldn't this be combined with the above?

Comment (manucorporat) [reply core/src/components/tabs/tabs.tsx]: not really notice that the one above resets this.useRouter

**Comment by DavidStrausz** [comment]
@manucorporat 
This broke ion-radio-group/ion-radio, they end up in an infinite-loop when reused.

**Comment by manucorporat** [comment]
@DavidStrausz can you open an issue explaining how to reproduce?

**Comment by DavidStrausz** [comment]
@manucorporat I can but this will take some time, just thought you might want a heads up :)

**Pull request merged**